@here/harp-geometry/lib/SubdivisionModifier.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+    Face3,
+    Geometry,
+} from "three";
+
+/**
+ * The `SubdivisionModifier` subdivides triangle mesh geometries positioned
+ * on the surface of a sphere centered at `(0, 0, 0)`.
+ */
+export class SubdivisionModifier {
+    /**
+     * Constructs a new `SubdivisionModifier`.
+     *
+     * @param angle The maximum angle in radians between two vertices and the origin.
+     */
+    constructor(readonly angle: number) { }
+
+    /**
+     * Subdivides the faces of the given `THREE.Geometry`.
+     *
+     * @param geometry The `THREE.Geometry` to subdivide.
+     */
+    modify(geometry: Geometry): Geometry {
+        const { vertices, faces } = geometry;
+
+        const cache = new Map<string, number>();
+
+        function middleVertex(i: number, j: number): number {
+            const key = `${Math.min(i, j)}_${Math.max(i, j)}`;
+
+            const h = cache.get(key);
+
+            if (h !== undefined) {
+                return h;
+            }
+
+            const index = vertices.length;
+
+            vertices.push(vertices[i].clone().add(vertices[j]).multiplyScalar(0.5));
+
+            cache.set(key, index);
+
+            return index;
+        }
+
+        // The resulting triangles.
+        const subdividedFaces: Face3[] = [];
+
+        while (true) {
+            const face = faces.shift();
+
+            if (!face) {
+                break;
+            }
+
+            // get the vertices of the current triangle.
+            const [a, b, c] = [vertices[face.a], vertices[face.b], vertices[face.c]];
+
+            // get the angles
+            const [alpha, beta, gamma] = [a.angleTo(b), b.angleTo(c), c.angleTo(a)];
+
+            // find the maximum angle
+            const m = Math.max(alpha, Math.max(beta, gamma));
+
+            // split the triangle if needed.
+            if (m <= this.angle) {
+                subdividedFaces.push(face);
+            } else if (alpha === m) {
+                const d = middleVertex(face.a, face.b);
+                faces.push(new Face3(face.a, d, face.c));
+                faces.push(new Face3(d, face.b, face.c));
+            } else if (beta === m) {
+                const d = middleVertex(face.b, face.c);
+                faces.push(new Face3(face.a, face.b, d));
+                faces.push(new Face3(face.a, d, face.c));
+            } else if (gamma === m) {
+                const d = middleVertex(face.c, face.a);
+                faces.push(new Face3(face.a, face.b, d));
+                faces.push(new Face3(d, face.b, face.c));
+            } else {
+                throw new Error("failed to subdivide the given geometry");
+            }
+        }
+
+        geometry.faces = subdividedFaces;
```

Comment:
Geometry may have some interdependencies between internal data.
Shouldn't we trigger one of these:
```
	elementsNeedUpdate;
	verticesNeedUpdate;
	this.uvsNeedUpdate =...
```
(https://threejs.org/docs/#api/en/core/Geometry)

Reply:
I updated the change to set vertices/elementsNeedUpdate. I also I added a paragraph in the documentation explaining that only the position and the faces of the geometry are modified, other common attributes of THREE.Geometry (like normals and UVs) are usually inferred by the vertex position so currently we don't need to preserve them, but if needed we can easily extend the code compute values for them.